autopilot/agent.go

```diff
@@ -430,232 +415,241 @@ func (a *Agent) controller() {
-					"need for more channels")
 			}
 
-			pendingMtx.Lock()
-			log.Debugf("Pending channels: %v", spew.Sdump(pendingOpens))
-			pendingMtx.Unlock()
-
-			// With all the updates applied, we'll obtain a set of
-			// the current active channels (confirmed channels),
-			// and also factor in our set of unconfirmed channels.
-			confirmedChans := a.chanState
-			pendingMtx.Lock()
-			totalChans := mergeChanState(pendingOpens, confirmedChans)
-			pendingMtx.Unlock()
-
-			// Now that we've updated our internal state, we'll
-			// consult our channel attachment heuristic to
-			// determine if we should open up any additional
-			// channels or modify existing channels.
-			availableFunds, numChans, needMore := a.cfg.Heuristic.NeedMoreChans(
-				totalChans, a.totalBalance,
-			)
-			if !needMore {
-				continue
-			}
+		// A new channel has been opened by the agent or an external
+		// subsystem, but is still pending confirmation.
+		case <-a.pendingOpenUpdates:
+			updateBalance()
 
-			log.Infof("Triggering attachment directive dispatch, "+
-				"total_funds=%v", a.totalBalance)
-
-			// We're to attempt an attachment so we'll o obtain the
-			// set of nodes that we currently have channels with so
-			// we avoid duplicate edges.
-			connectedNodes := a.chanState.ConnectedNodes()
-			pendingMtx.Lock()
-			nodesToSkip := mergeNodeMaps(connectedNodes, failedNodes, pendingOpens)
-			pendingMtx.Unlock()
-
-			// If we reach this point, then according to our
-			// heuristic we should modify our channel state to tend
-			// towards what it determines to the optimal state. So
-			// we'll call Select to get a fresh batch of attachment
-			// directives, passing in the amount of funds available
-			// for us to use.
-			chanCandidates, err := a.cfg.Heuristic.Select(
-				a.cfg.Self, a.cfg.Graph, availableFunds,
-				numChans, nodesToSkip,
-			)
-			if err != nil {
-				log.Errorf("Unable to select candidates for "+
-					"attachment: %v", err)
-				continue
-			}
+		// The balance of the backing wallet has changed, if more funds
+		// are now available, we may attempt to open up an additional
+		// channel, or splice in funds to an existing one.
+		case <-a.balanceUpdates:
+			log.Debug("Applying external balance state update")
+
+			updateBalance()
+
+		// The channel we tried to open previously failed for whatever
+		// reason.
+		case <-a.chanOpenFailures:
+			log.Debug("Retrying after previous channel open " +
+				"failure.")
 
-			if len(chanCandidates) == 0 {
-				log.Infof("No eligible candidates to connect to")
+			updateBalance()
+
+		// New nodes have been added to the graph or their node
+		// announcements have been updated. We will consider opening
+		// channels to these nodes if we haven't stabilized.
+		case <-a.nodeUpdates:
+			log.Infof("Node updates received, assessing " +
+				"need for more channels")
+
+		// The agent has been signalled to exit, so we'll bail out
+		// immediately.
+		case <-a.quit:
+			return
+		}
```

Comment:
I like this idea, as querying the heuristic has proven to be resource intensive and it will help ensure we gather as many triggers in each attempt. We could consider adding a timeout after processing the first update, to catch events that arrive in spurts, but perhaps with slight gaps if relying on non-blocking select to determine completeness
```golang
for {
    // blocks until one update comes thru
    needsBalanceUpdate := waitForNextUpdate() 
    // same as prior, but with for loop + timeout to catch stragglers
    needsBalanceUpdate |= drainUpdates(heuristicQueryDelay) 

    ... query heuristic, etc
}
```

Reply:
Great suggestion! I don't think it's worth blocking this PR to implement this feedback, but as we start to really dig into the `autopilot` package over the next few weeks, it's definitely something we should re-examine, I've made a note to come back to this in the future.